### PR TITLE
Course api

### PR DESCRIPTION
**Pull request by PoornimaShridhar**
course api added

### CODE REVIEW

File: course/api.py [hunk 1]
```diff
@@ -33,6 +38,75 @@ class CourseViewSet(viewsets.ModelViewSet):
...
+                }
+                return Response(data, status.HTTP_400_BAD_REQUEST)
+        return True
+
+    def create(self, request):
+        user = request.user
+        data = request.data
+        if not has_valid_subscription(user):
+            data = {
+                "error": "User: {} does not have a valid subscription or the"
```
Comment (suraj-iitb) [inline course/api.py]: Use multiline string

File: course/api.py [hunk 2]
```diff
@@ -33,6 +38,75 @@ class CourseViewSet(viewsets.ModelViewSet):
...
+                "limit of number of courses in subscription exceeded".format(user),
+            }
+            return Response(data, status.HTTP_401_UNAUTHORIZED)
```
Comment (suraj-iitb) [inline course/api.py]: Use 403 forbiden

File: course/api.py [hunk 3]
```diff
@@ -33,6 +38,75 @@ class CourseViewSet(viewsets.ModelViewSet):
...
+                    "error": "User: {} not registered in course with id: {}".format(
+                        user, course_id
+                    ),
+                }
+                return Response(data, status.HTTP_400_BAD_REQUEST)
+        return True
+
+    def create(self, request):
+        user = request.user
+        data = request.data
```
Comment (suraj-iitb) [inline course/api.py:71]: add a enter after line 71

File: course/api.py (hunk 2 above)
Comment (suraj-iitb) [reply course/api.py]: add a enter after line 77

File: course/api.py [hunk 4]
```diff
@@ -33,6 +38,75 @@ class CourseViewSet(viewsets.ModelViewSet):
...
+            }
+            return Response(data, status.HTTP_401_UNAUTHORIZED)
+        serializer = self.get_serializer(data=data)
+        if serializer.is_valid(raise_exception=True):
+            serializer.save()
+            course_history = CourseHistory(
+                user=user, course=serializer.instance, role="I", status="E"
+            )
+            course_history.save()
+            return Response({}, status=status.HTTP_201_CREATED)
```
Comment (suraj-iitb) [inline course/api.py]: return serializer.data

File: course/api.py [hunk 5]
```diff
@@ -33,6 +38,75 @@ class CourseViewSet(viewsets.ModelViewSet):
     )
     ordering_fields = ("id",)
 
+    def _checks(self, course_id, user):
+        try:
+            CourseHistory.objects.filter(course_id=course_id, user=user)
```
Comment (suraj-iitb) [inline course/api.py]: This check already in the permissions check of has_object_permission. remove  it

File: course/api.py [hunk 6]
```diff
@@ -33,6 +38,75 @@ class CourseViewSet(viewsets.ModelViewSet):
...
+        except CourseHistory.DoesNotExist:
+            data = {
+                "error": "Course history of user with id: {} does not exist".format(
+                    user
+                ),
+            }
+            return Response(data, status.HTTP_400_BAD_REQUEST)
+        if Course.objects.filter(
+            id=course_id, owner=user
+        ) and not has_valid_subscription(user):
```
Comment (suraj-iitb) [inline course/api.py]: course. owner instead of user in has_valid_subscription(user)

File: course/api.py [hunk 7]
```diff
@@ -33,6 +38,75 @@ class CourseViewSet(viewsets.ModelViewSet):
...
+            data = {
+                "error": "User: {} does not have a valid subscription or the"
+                "limit of number of courses in subscription exceeded".format(user),
+            }
+            return Response(data, status.HTTP_401_UNAUTHORIZED)
+        elif CourseHistory.objects.filter(Q(role="I") | Q(role="T")):
+            if not check_course_registration(course_id, user):
```
Comment (suraj-iitb) [inline course/api.py:60]: You can directly check for is_registered in the course.

File: course/api.py [hunk 8]
```diff
@@ -33,6 +38,75 @@ class CourseViewSet(viewsets.ModelViewSet):
...
+                data = {
+                    "error": "User: {} not registered in course with id: {}".format(
+                        user, course_id
+                    ),
+                }
+                return Response(data, status.HTTP_400_BAD_REQUEST)
```
Comment (suraj-iitb) [inline course/api.py]: HTTP_401_UNAUTHORIZED if not registered

File: course/api.py (hunk 1 above)
Comment (suraj-iitb) [reply course/api.py]: also create and _check have some common code for has_subscription check. write a separate fun for it

File: course/api.py [hunk 9]
```diff
@@ -33,6 +38,75 @@ class CourseViewSet(viewsets.ModelViewSet):
...
+    @action(detail=True, methods=["PUT", "PATCH"])
+    def update_course(self, request, pk):
+        user = request.user
+        data = request.data
+        check = self._checks(pk, user)
+        if check:
+            serializer = self.get_serializer(self.get_object(), data=data, partial=True)
+            if serializer.is_valid(raise_exception=True):
+                serializer.save()
+                print("student")
```
Comment (suraj-iitb) [inline course/api.py]: remove print

File: course/api.py [hunk 10]
```diff
@@ -33,6 +38,75 @@ class CourseViewSet(viewsets.ModelViewSet):
...
+                return Response(serializer.data)
+        return check
+
+    @action(detail=True, methods=["DELETE"], permission_classes=[Isowner])
+    def delete_course(self, request, pk):
+        instance = self.get_object()
+        instance.delete()
+        return Response(status=status.HTTP_204_NO_CONTENT)
+        data = {
```
Comment (suraj-iitb) [inline course/api.py]: line 105 to 108 is not required

File: course/serializers.py [hunk 11]
```diff
@@ -1,20 +1,49 @@
...
+
     class Meta:
         model = Course
         fields = "__all__"
 
     def create(self, validated_data):
-        instance = self.Meta.model(**validated_data)
+        df_settings_data = validated_data.pop("df_settings")
+        course = Course.objects.create(**validated_data)
+        # for df_setting_data in df_settings_data:
```
Comment (suraj-iitb) [inline course/serializers.py]: remove this comment

File: course/serializers.py [hunk 12]
```diff
@@ -1,20 +1,49 @@
...
+        # for df_boolean_field in df_boolean_fields:
+        #     if df_boolean_field not in df_settings_data_keys:
+        #         df_settings_data[df_boolean_field]=False
+        # validated_data_keys = validated_data.keys() if validated_data else []
+        # course_boolean_fields = ["is_published"]
+        # for course_boolean_field in course_boolean_fields:
+        #     if course_boolean_field not in validated_data_keys:
+        #         validated_data[course_boolean_field]=False
+        # ends here
+        # info = model_meta.get_field_info(instance)
```
Comment (suraj-iitb) [inline course/serializers.py]: delete this line

File: course/serializers.py [hunk 13]
```diff
@@ -1,20 +1,49 @@
...
+        for attr, value in validated_data.items():
+            setattr(instance, attr, value)
         instance.save()
-        course_history = CourseHistory(
-            user=validated_data["owner"], course=instance, role="I", status="E"
-        )
-        course_history.save()
+        discussion_forum = DiscussionForum.objects.get(course=instance)
```
Comment (suraj-iitb) [inline course/serializers.py]: use get_object_or_404 to handle exception as get is unsafe

File: course/tests.py [hunk 14]
```diff
@@ -32,83 +69,166 @@ def test_get_course(self):
...
     def test_create_course(self):
         """
         Ensure we can create a new Course object.
         """
-        self.client.login(email="test1@test.com", password="Test@1001")
+        self.login(**CourseViewSetTest.ins_cred)
+        self.create_course_helper(status.HTTP_201_CREATED, "Course 1")
+        self.logout()
+        self.login(**CourseViewSetTest.ta_cred)
+        self.create_course_helper(status.HTTP_401_UNAUTHORIZED, "Course 2")
```
Comment (suraj-iitb) [inline course/tests.py]: Once u use 403 forbidden in create, use same code here

File: course/tests.py [hunk 15]
```diff
@@ -32,83 +69,166 @@ def test_get_course(self):
...
+        self.logout()
+        self.login(**CourseViewSetTest.stu_cred)
+        self.create_course_helper(status.HTTP_401_UNAUTHORIZED, "Course 3")
```
Comment (suraj-iitb) [inline course/tests.py]: 403

File: course/tests.py [hunk 16]
```diff
@@ -32,83 +69,166 @@ def test_get_course(self):
...
+        self.assertEqual(response.status_code, status_code)
 
     def test_partial_update_course(self):
         """
         Ensure we can partially update an existing Course object.
         """
-        self.client.login(email="test1@test.com", password="Test@1001")
-        course = Course(
-            owner_id=CourseViewSetTest.user.id, title="Course6", course_type="O"
+        course = Course(owner=CourseViewSetTest.ins, title="Course6", course_type="O")
```
Comment (suraj-iitb) [inline course/tests.py:171]: maybe try to write function for common code (update & partial_update): Course and Df Creation

File: utils/drf_utils.py [hunk 17]
```diff
@@ -183,7 +184,8 @@ def has_object_permission(self, request, view, obj):
             course_histories = CourseHistory.objects.filter(
                 Q(course=obj) & (Q(role="I") | Q(role="T")) & Q(user=request.user)
             )
-            if course_histories is not None:
+            print(course_histories)
```
Comment (suraj-iitb) [inline utils/drf_utils.py]: remove print

File: utils/drf_utils.py [hunk 18]
```diff
@@ -300,3 +302,33 @@ def has_permission(self, request, view):
...
+        return False
+
+    def has_object_permission(self, request, view, obj):
+        """
+        Applicable at model instance level (retrieve, update, partial_update, destroy)
+        """
+        if request.user.is_authenticated:
+            if request.method in permissions.SAFE_METHODS:
+                return True
+            course = Course.objects.filter(Q(id=obj.id) & Q(owner=request.user))
```
Comment (suraj-iitb) [inline utils/drf_utils.py]: directly check obj.owner == request.user

File: utils/utils.py [hunk 19]
```diff
@@ -44,3 +45,21 @@ def is_instructor_or_ta(course_id, user):
...
+    subscription_history = SubscriptionHistory.objects.filter(user=user)
+    if (
+        subscription_history
+        and len(no_of_courses) < subscription_history[0].subscription.no_of_courses
+    ):
+        return True
+    return False
+
+
+def is_owner(course_id, user):
```
Comment (suraj-iitb) [inline utils/utils.py]: delete this

File: course/api.py (hunk 7 above)
Comment (PoornimaShridhar) [reply course/api.py:60]: is_registered is not an attribute in course/courseHistory model

File: course/api.py (hunk 3 above)
Comment (PoornimaShridhar) [reply course/api.py:71]: All these are part of the same function. Enter line would break it

**Review by suraj-iitb — APPROVED** [review]

**Pull request merged**